contracts/CollateralProvider/CollateralProvider.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../ProviderInterface/IProviderSingleIdRegistrar.sol";
+import "../Provider/ProviderState.sol";
+import "./CollateralState.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+
+contract CollateralProvider is
+    IProviderSingleIdRegistrar,
+    CollateralState,
+    ProviderState,
+    IERC721Receiver
+{
+    ///@dev withdraw tokens
+    function onERC721Received(
+        address operator,
+        address from,
+        uint256 poolId,
+        bytes calldata
+    ) external override returns (bytes4) {
+        require(msg.sender == address(lockDealNFT), "invalid nft contract");
+        if (operator == from) {
+            uint256 tokenPoolId = poolId + 1;
+            DealProvider dealProvider = dealProvider.dealProvider();
+            // create new pool with tokens
+            uint256 newPoolId = lockDealNFT.mintForProvider(from, address(dealProvider));
+            uint256[] memory params = new uint256[](1);
+            params[0] = dealProvider.getParams(tokenPoolId)[0];
+            dealProvider.registerPool(newPoolId, params);
+
+            // return nft to owner
+            lockDealNFT.transferFrom(address(this), from, poolId);
```

Comment:
To make sure, this will not trigger the onERC721Received, in case of contract as the user

Reply:
need to add a re-entrance guard here, it can be looped